feat: ✨ bring in content on commits and PRs from elsewhere (#157)
seedcase-project/guidebook: workflows/prs.qmd

## Creating a pull request

Before making changes, create a short-lived branch from an up-to-date
`main` branch. Keep the branch focused on one issue or logical change so

> that its PR will also be focused.

Reviewer: Related to this; should this PR be two PRs? One for commits and one for PRs? 🤔

### Branch naming conventions

Branch names should be lowercase, use hyphens between words, and follow
the
[Conventional Branch](https://github.com/pshaddel/conventional-branch)

> format `{type}/{short-description}`. Write the description in the

Reviewer: This is not the default format for the extension (we've specified this format in the extension settings ourselves), so maybe add the extension as a tip below instead.

```suggestion
format `{type}/{short-description}`. Write the description in the
```

imperative. The `type` is based on the
[Conventional Commits](commits.qmd#conventional-commits) types (e.g.,
`docs`, `fix`, `feat`, `refactor`). For example:

- `docs/add-pr-workflow-guide`
- `fix/broken-link-in-readme`
- `refactor/update-template-structure`

The
[Conventional Branches VS Code extension](https://marketplace.visualstudio.com/items?itemName=pshaddel.conventional-branch)
can help create a correctly formatted branch name from the Command

> Palette.

Reviewer: ```suggestion
Palette. In our templates, e.g., [`template-python-package`](https://template-python-package.seedcase-project.org/), we've added the VS Code settings for the extension to follow the format described above.
```

### Prepare the change

> Start with an issue before investing substantial time in an

Reviewer: ```suggestion
Start with creating an issue (or comment on an existing one) before investing substantial time on an
```

implementation. Use the issue to describe the problem, expected
behaviour, and proposed direction. This allows the team to agree on the
scope and approach before work begins so that effort is not wasted on
substantial re-writes or even PR rejection.

> Before requesting a review:

Reviewer: ```suggestion
Before creating a PR:
```

- Review the complete diff yourself, not only the most recent changes.
- Remove accidental or unrelated changes.
- Run the repository's automated formatting, linting, tests, and other
  checks.
- Check that code and text communicate their intent without relying on

>   the PR discussion.

Reviewer: ```suggestion
  the PR description.
```

the format `{type}/{short-description}`. Write the description in the
imperative. The `type` is based on the
[Conventional Commits](commits.qmd#conventional-commits) types (e.g.,
`docs`, `fix`, `feat`, `refactor`). For example:

- `docs/add-pr-workflow-guide`
- `fix/broken-link-in-readme`
- `refactor/update-template-structure`

The
[Conventional Branches VS Code extension](https://marketplace.visualstudio.com/items?itemName=pshaddel.conventional-branch)
can help create a correctly formatted branch name from the Command
Palette. In our templates, e.g.,
[`template-python-package`](https://template-python-package.seedcase-project.org/),
we've added the VS Code settings for the extension to follow the format
described above.

### Prepare the change

Start with creating an issue (or comment on an existing one) before
investing substantial time on an implementation. Use the issue to
describe the problem, expected behaviour, and proposed direction. This
allows the team to agree on the scope and approach before work begins so
that effort is not wasted on substantial re-writes or even PR rejection.

Before creating a PR:

> - Review the complete diff yourself, not only the most recent changes.

Reviewer 2: ```suggestion
- Review all the changes made yourself, not only the most recent changes.
```

- Remove accidental or unrelated changes.
- Run the repository's automated formatting, linting, tests, and other

>   checks.

Reviewer 2: ```suggestion
  checks (found in the `justfile`).
```

### When to create a PR

You can create a PR as a draft once you have made meaningful changes and
have a clear idea of the implementation. An early draft makes the work
visible and lets you ask for early feedback if needed before committing
too much time to a specific approach. Mark it as a

> [draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft)

Reviewer: Hnnnm, do we actually follow this workflow? I've rarely seen or made draft PRs that are "in review" on the board?

Author: True, I see it as an optional step, but I currently only remember myself putting up a draft PR and asking for early feedback is myself =p I do generally think the idea of showing what we are working on is useful though and just because we don't do this often, doesn't mean we can't have it as an optional step. However, if you think its a bad idea to do this, we can of course remove it.

Reviewer 2: I know @martonvago does this fairly often, is mark a PR as draft while working on the PR. Though usually I notice it is in draft state for not too long.

> while it is incomplete, and state what feedback would be useful.

Reviewer: ```suggestion
while it is incomplete, and add to the PR description what feedback would be useful.
```

> Only mark the PR ready and request a full review after the work is

Reviewer: ```suggestion
Only mark the PR ready and request a review after the work is
```

quicker to review, easier to test, and simpler to revert if something
goes wrong.

As a practical guide for code, aim for no more than a few hundred
changed lines. Around 400 changed lines can be a useful upper threshold
of when to reconsider the scope of the PR, but it is not a hard limit: generated
files and a logically indivisible change may justify a larger diff while
logically separate ideas require splitting even if they change just a
few lines of code. Explain why a large PR cannot reasonably be split.

Avoid stacked PRs, where one open PR targets the branch of another open
PR, unless there is an exceptional reason to use them. Dependencies
between PRs make review and merging harder. We prefer independent changes;

> otherwise, clearly identify the dependency and required merge order.

Author: We might reconsider this now that the stacked PR feature on github is evlolving, but I left it for now

quicker to review, easier to test, and simpler to revert if something
goes wrong.

> As a practical guide for code, aim for no more than a few hundred

Reviewer: ```suggestion
As a practical guide for code, aim for around 100
```

I like this description of "small" from Google's [eng-practices](https://google.github.io/eng-practices/review/developer/small-cls.html) (they call PR's "CL" for "changelist"):

> There are no hard and fast rules about how large is “too large.” 100 lines is usually a reasonable size for a CL, and 1000 lines is usually too large, but it’s up to the judgment of your reviewer. The number of files that a change is spread across also affects its “size.” A 200-line change in one file might be okay, but spread across 50 files it would usually be too large.

> Keep in mind that although you have been intimately involved with your code from the moment you started to write it, the reviewer often has no context. What seems like an acceptably-sized CL to you might be overwhelming to your reviewer. When in doubt, write CLs that are smaller than you think you need to write. Reviewers rarely complain about getting CLs that are too small.

Author: Nice find, I tried to incorporate most of it in the latest version.

Reviewer: Very nice! You could even add a link to it? If you don't agree, leave as is :)

changed lines. Around 400 changed lines can be a useful upper threshold

> of when to reconsider the scope, but it is not a hard limit: generated

Reviewer: ```suggestion
of when to reconsider the scope of the PR, but it is not a hard limit: generated
```

files and a logically indivisible change may justify a larger diff while
logically separate ideas require splitting even if they change just a
few lines of code. Explain why a large PR cannot reasonably be split.

Avoid stacked PRs, where one open PR targets the branch of another open
PR, unless there is an exceptional reason to use them. Dependencies

> between PRs make review and merging harder. Prefer independent changes;

Reviewer: ```suggestion
between PRs make review and merging harder. We prefer independent changes;
```

otherwise, clearly identify the dependency and required merge order.

### Writing a good PR description

A good PR description helps reviewers understand the context and purpose
of the changes. Our repositories include a
[pull request template](https://github.com/seedcase-project/guidebook/blob/main/.github/pull_request_template.md)

> to make it easy to fill out the description when creating a PR. In the

Reviewer: ```suggestion
to make it easier to follow the same description structure when creating a PR. In the
```

> to make it easy to fill out the description when creating a PR. In the

Reviewer: You could also add this as a tip at the end of the section instead, saying that our PR template structure follows the practices described above.

> title, write a [Conventional Commit](commits.qmd#conventional-commits),

Reviewer: ```suggestion
title, follow the [Conventional Commit](commits.qmd#conventional-commits) convention,
```

> for example `fix: prevent overwriting raw files`. Because we squash PRs,

Reviewer: This seems like a good place for the PR title section that's currently in `commits.qmd` after L164.

Author: I think a version of that content already existed elsewhere in this file; at the end under "berfore mergeing"

> this title normally becomes the commit subject on `main`.

Reviewer: ```suggestion
this title becomes the commit subject on `main`. Therefore, it's important that you ensure, before you merge a PR, that the title follow the conventional commits convention. Otherwise, the changelog and potential version update won't be auto-generated.
```

- **Why**: Explain the motivation or reasoning behind the changes.
- **How**: If the changes are not straightforward, describe the approach
  you took.
- **Evidence**: Include tests, usage examples, or before-and-after
  screenshots when they help demonstrate the changed behaviour.

If the PR addresses an existing issue, reference it by writing
`Closes #<issue-number>`, e.g., `Closes #2` in the description. This
automatically links the PR to the issue and closes the issue when the PR
is merged.

Keep the description **concise but informative**, with detail
proportional to the complexity of the change. A few sentences are often
enough for a small PR. A deep or unusually complex change may need a
longer explanation of constraints and design choices so that the
reviewer does not have to reconstruct the author's reasoning from the
diff. Writing informative commit messages will help the reviewer if they
need additional detail about a specific change in the PR.

Do not duplicate a full bug report or design discussion from an issue.
Link to it and summarise only the context needed to review the concrete
implementation. Keep general questions and alternative future work in

> targeted issues rather than expanding the scope of the PR discussion.

Reviewer: We also sometimes tend to add comments to our own PRs to address questions or considerations we've come across in our work. Do we want to address that here (and allow that as good practice in general?)?

Author: Great point! I added a bullet point called "considerations" above to try to capture this.

quicker to review, easier to test, and simpler to revert if something
goes wrong.

There is no strict guideline for how big a single PR can be in terms of
the lines of code changed; the most important is to keep logically
connected changes together in the same PR. Generated files and a

> logically indivisible change may justify a larger diff while logically

Reviewer 2: ```suggestion
logically indivisible change may justify a larger change while logically
```

separate ideas require splitting even if they change just a few lines of
code. Having that said, a practical recommendation is to start asking
yourself whether the PR could be broken up into a few smaller PRs when
you approach 100-150 lines of code. Keep in mind that your reviewer does

> not have the same context as you do, and they have not been seen the

Reviewer 2: ```suggestion
not have the same context as you do, and they have not seen the
```

changes gradually progress as you have, so smaller PRs makes it easier
for them to understand what is going on. If you think a large PR should
not be split, explain why in the PR description, so this is clear to the
reviewer. More guidelines on how to write small PRs could be found in
e.g.
[Google's engineering pracitces](https://google.github.io/eng-practices/review/developer/small-cls.html).

Avoid stacked PRs, where one open PR targets the branch of another open
PR, unless there is an exceptional reason to use them. Dependencies
between PRs make review and merging harder. We prefer independent
changes; otherwise, clearly identify the dependency and required merge

> order.

Reviewer 2: ```suggestion
order.

<!-- TODO: May need to update the advice on stacked PRs with GitHub's new feature -->
```

### Writing a good PR description

A good PR description helps reviewers understand the context and purpose
of the changes. Our repositories include a
[pull request template](https://github.com/seedcase-project/guidebook/blob/main/.github/pull_request_template.md)
to make it easier to follow the same description structure when creating
a PR. In the title, follow the

> [Conventional Commit](commits.qmd#conventional-commits) convention, for

Reviewer 2: ```suggestion
[Conventional Commit](commits.qmd#conventional-commits) format, for
```

example `fix: prevent overwriting raw files`. Because we squash PRs,
this title becomes the commit subject on `main`. Therefore, it's
important that you ensure, before you merge a PR, that the title follow

> the conventional commits convention. Otherwise, the changelog and

Reviewer 2: ```suggestion
Conventional Commits. Otherwise, the changelog and
```

potential version update won't be auto-generated.

In the description field, aim to cover:

- **What**: Briefly describe the changes you made.
- **Why**: Explain the motivation or reasoning behind the changes.
- **How**: If the changes are not straightforward, describe the approach
  you took.
- **Evidence**: Include tests, usage examples, or before-and-after
  screenshots when they help demonstrate the changed behaviour.
- **Considerations**: Describe why alternative approaches were rejected,

>   try to address concerns/confusions you might foresee, raise

Reviewer 2: ```suggestion
  try to address concerns/confusions you might foresee, or raise
```

  outstanding questions to clearly define what is yet to be decided on,

>   etc.

Reviewer 2: ```suggestion
  outstanding questions to clearly define what is yet to be decided on.
```

If the PR addresses an existing issue, reference it by writing
`Closes #<issue-number>`, e.g., `Closes #2` in the description. This
automatically links the PR to the issue and closes the issue when the PR
is merged.

Keep the description **concise but informative**, with detail
proportional to the complexity of the change. A few sentences are often
enough for a small PR. A deep or unusually complex change may need a
longer explanation of constraints and design choices so that the
reviewer does not have to reconstruct the author's reasoning from the

> diff. Writing informative commit messages will help the reviewer if they

Reviewer 2: ```suggestion
PR. Writing informative commit messages will help the reviewer if they
```

quicker to review, easier to test, and simpler to revert if something
goes wrong.

There is no strict guideline for how big a single PR can be in terms of
the lines of code changed; the most important is to keep logically
connected changes together in the same PR. Generated files and a
logically indivisible change may justify a larger diff while logically
separate ideas require splitting even if they change just a few lines of
code. Having that said, a practical recommendation is to start asking
yourself whether the PR could be broken up into a few smaller PRs when
you approach 100-150 lines of code. Keep in mind that your reviewer does
not have the same context as you do, and they have not been seen the
changes gradually progress as you have, so smaller PRs makes it easier
for them to understand what is going on. If you think a large PR should

> not be split, explain why.

Reviewer: ```suggestion
not be split, explain why in the PR description, so this is clear to the reviewer.
```

Setting both ensures that it is clear who owns the PR and who needs to
look at it. We've set it up so this is done automatically (see our
[template](https://template-website.seedcase-project.org/) for how).
After creating the PR, add it to the relevant GitHub Project so that the

> team can track the work.

Reviewer: ```suggestion
team can track the work. If you're still working on it, set it as "In progress". Once it's ready for review, set it to "In review", so it's clearly communicated to the team.
```

### Responding to reviews

Treat review as a collaboration to improve the change, not as an
obstacle to merging it. For each review comment:

> - Reply when you make the requested change; do not silently update the

Reviewer: ```suggestion
- Reply when you make the requested change, do not silently update the
```

  code or text.
- Ask for clarification rather than guessing when a comment is unclear.
- Explain respectfully when you disagree and focus on the work rather

>   than the person (avoid using "you" as much as possible).

Reviewer: I tend to write "we" so it feels more like we're collaborating and on the same team :grin:

Author: I like that too =)

Reviewer: Maybe we (😉) could add that as a separate list item or incorporate the use of "we" in this item?

- Be patient with conversations. Your reviewer might be incorrect or you
  might be misunderstanding something. The only way to find out is

>   through effective communication.

Reviewer: ```suggestion
- Be patient with conversations. You or your reviewer might be incorrect, or you
  might be misunderstanding each other. The only way to find out is
  through communication. Remember that you're working towards the same goal.
```
:grin:

- Resolve a conversation only after the concern has been addressed or
  everyone has agreed on the outcome.

After addressing the feedback, check the full diff and automated checks
again, then re-request a review. Be grateful when a review catches a
problem before it reaches `main`; an effective review can help you avoid

> all the headaches that would arise if one of your bugs made it into

Reviewer: ```suggestion
the headache that would arise if one of your bugs made it into
```

Setting both ensures that it is clear who owns the PR and who needs to
look at it. We've set it up so this is done automatically (see our
[template](https://template-website.seedcase-project.org/) for how).
After creating the PR, add it to the relevant GitHub Project so that the
team can track the work. If you're still working on it, set it as "In
progress". Once it's ready for review, set it to "In review", so it's
clearly communicated to the team.

### Responding to reviews

Treat review as a collaboration to improve the change, not as an
obstacle to merging it. For each review comment:

- Reply when you make the requested change, do not silently update the

>   code or text.

Reviewer 2: Hmm, I don't think this is necessary. Clicking "resolved" is enough (as the commits afterwards will show the change).

```suggestion
- When you make the requested change, click the "resolve" button on the comment 
  to indicate you addressed it.
```

- Ask for clarification rather than guessing when a comment is unclear.
- Explain respectfully when you disagree and focus on the work rather
  than the person. A practical tip here is to avoid using "you" in
  comments when it can create distance between yourself and the
  reviewer; either replacing it with "we" for a more collaborative tone
  or describing the issue more objectively without a personal pronoun
  altogether (this doesn't mean it's never good to write "you", use your
  best judgment and think about how a reviewer might interpret your
  comment differently from your own interpretation).
- Be patient with conversations. You or your reviewer might be
  incorrect, or you might be misunderstanding each other. The only way
  to find out is through communication. Remember that you're working
  towards the same goal.
- Resolve a conversation only after the concern has been addressed or
  everyone has agreed on the outcome.

> After addressing the feedback, check the full diff and automated checks

Reviewer 2: ```suggestion
After addressing the feedback, review all the changes in the PR and automated checks
```

again, then re-request a review. Be grateful when a review catches a
problem before it reaches `main`; an effective review can help you avoid
the headache that would arise if one of your bugs made it into

> production.

Reviewer 2: ```suggestion
the headache that would arise if one of your bugs was found in a release.
```

PRs are tasks of the highest priority. Timely reviews keep the work
moving and prevent bottlenecks.

### Reviewing changes

> Review promptly, ideally within one to three working days. A slow review

Reviewer: ```suggestion
Review as promptly as possible, ideally within one to three working days. A slow review process
```

causes the author and reviewer to lose context, increases the chance of

> merge conflicts, and delays the rest of the team. If you cannot review

Reviewer: ```suggestion
merge conflicts, and might delay the rest of the team. If you cannot review
```

> within that time, tell the author so that another reviewer can help.

Reviewer: ```suggestion
within that time, let the author know (preferable in a PR comment or, alternatively, at an update meeting) so another reviewer can help.
```

> Start by reading the linked issue and PR description. Then review the

Reviewer: ```suggestion
Start by reading the linked issue and PR description. Then, review the
```

complete diff and defer to individual commit messages to understand why
a change was introduced. If it is hard to understand what the code does
without reading the commit message, suggest that the code is rewritten
or an explanatory comment is added. Focus on correctness, clarity,
maintainability, tests, documentation, and whether the change stays
within the agreed scope. Automated checks should handle routine
formatting and syntax issues wherever possible, leaving people to focus
on decisions that require judgement.

When leaving comments, be specific about what you think should change
and why. Avoid vague feedback like "this doesn't look right" and instead
explain what the issue is and how it could be improved so that your
feedback is both constructive and actionable. Explain both the concern
and why it matters.

Comment on the work, not the author, and remain open to the possibility
that you have misunderstood the change. If code is hard to understand
without a commit message or long discussion, suggest making the code

> clearer or adding a concise explanatory comment.

Reviewer: ```suggestion
that you have misunderstood the change.
```
This is a bit repetitive of L317-324, so we can remove it here.

For Markdown and Quarto files, use the "Display the rich diff" button in
the "Files changed" tab to review the rendered content as well as the

> source.

Reviewer: What a minute; where's this button? 😅

Author: I think it only exist for markdown so I removed the "quarto" mention

PRs are tasks of the highest priority. Timely reviews keep the work
moving and prevent bottlenecks.

### Reviewing changes

Review as promptly as possible, ideally within one to three working
days. A slow review process causes the author and reviewer to lose
context, increases the chance of merge conflicts, and might delay the
rest of the team. If you cannot review within that time, let the author
know (preferable in a PR comment or, alternatively, at an update
meeting) so another reviewer can help.

Start by reading the linked issue and PR description. Then, review the

> complete diff and defer to individual commit messages to understand why

Reviewer 2: ```suggestion
changes and defer to individual commit messages to understand why
```

a change was introduced. If it is hard to understand what the code does
without reading the commit message, suggest that the code is rewritten
or an explanatory comment is added. Focus on correctness, clarity,
maintainability, tests, documentation, and whether the change stays
within the agreed scope. Automated checks should handle routine
formatting and syntax issues wherever possible, leaving people to focus
on decisions that require judgement.

When leaving comments, be specific about what you think should change
and why. Avoid vague feedback like "this doesn't look right" and instead
explain what the issue is and how it could be improved so that your
feedback is both constructive and actionable. Explain both the concern
and why it matters.

Comment on the work, not the author, and remain open to the possibility
that you have misunderstood the change.

For Markdown files, use the "Display the rich diff" button in the "Files
changed" tab to review the rendered content as well as the source.

### Making suggestions

> Use GitHub's suggestion feature to propose specific actionable changes

Reviewer 2: ```suggestion
Use GitHub's suggestion feature in the comment textbox to propose specific actionable changes
```

"Re-request review" button in the right sidebar of the PR page. This
back and forth continues until the PR is approved and ready to merge.

> Keep this discussion sharply focused on the PR. If a small part prompts

Reviewer: ```suggestion
Keep the discussion sharply focused on the PR. If a small part prompts
```

an extensive design discussion, create a targeted issue and continue the
broader conversation there. More than about 50 comments or 25 commits is

> a useful signal that a PR may need to be split, although logical scope

Reviewer: ```suggestion
a useful signal that a PR may need to be split, although the logical scope
```

**"Squash and merge"**, which combines all the commits in the PR into a
single commit on `main`. This keeps the commit history clean and focused
on the overall change rather than individual commits.

> Before merging:

Reviewer: Could this be its own section? It doesn't entirely feel right under "dealing with merge conflicts"? 

```suggestion
### Before merging
```

**"Squash and merge"**, which combines all the commits in the PR into a
single commit on `main`. This keeps the commit history clean and focused
on the overall change rather than individual commits.

### Before merging

- Confirm that the required reviews are approved and automated checks
  pass.
- Confirm that the PR branch has no unresolved conflicts with `main`.
- Check that the squash commit title follows Conventional Commits.
- Edit the generated squash commit body into a durable explanation of
  the final change, removing temporary review notes and template text.
- Retain references to issues that should close when the PR is merged.

Delete the feature branch after merging unless it is still needed for a

> specific reason.

Reviewer 2: ```suggestion
specific reason.

::: callout-tip
Many of the above items can be automated within GitHub's settings.
For example, there is a setting to delete branches automatically
after it's merged.
:::
```
